File: Python/Data Structures/Queue/circular_queue_array.py
# Implementation of circular queue on a array
class CircularQueue:
    # initialize front and rear
    def __init__(self, capicity: int) -> None:
        self.front = -1
        self.rear = -1
        self.capicity = capicity
        self.queue = [None] * self.capicity
        self.size = 0

    def enqueue(self, val):
        if self.front == -1 and self.rear == -1:
            self.front = 0
            self.rear = 0
            self.queue[self.rear] = val
            self.size += 1

        elif (self.rear + 1) % self.capicity == self.front:
            print("Queue is full.")
            raise OverflowError

        else:
            self.rear = (self.rear + 1) % self.capicity
            self.queue[self.rear] = val
            self.size += 1

        print(f"Enqueued: {val}")

    def dequeue(self):
        if self.front == -1 and self.rear == -1:
            print("Queue is empty!")
        elif self.front == 0 and self.rear == 0:
            print(f"Dequeued: {self.front}")
            self.queue[self.front] = None
            self.front = -1
            self.rear = -1
            self.size -= 1
        else:
            print(f"Dequeued: {self.queue[self.front]}")
            self.queue[self.front] = None
            self.front = self.front + 1 % self.capicity
            self.size -= 1

    def display(self):
        if self.rear == -1 and self.front == -1:
            print("Queue is empty")
            return

        temp = self.front
        items = []
        for _ in range(self.size):
            items.append(self.queue[temp])
            temp = (temp + 1) % self.capicity

        print("Elements in the list: ", items)


if __name__ == "__main__":
    q = CircularQueue(5)

    q.enqueue(10)
    q.enqueue(20)
    q.enqueue(30)
    q.enqueue(40)
    q.enqueue(50)
    q.display()

    q.dequeue()
    q.dequeue()
    q.display()

    q.enqueue(60)
    q.enqueue(70)
    q.display()

File: Practice/Queue/circular_queue_array.c
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>

#define MAX 10

// define queue
typedef struct queue {
  int front;
  int rear;
  int queue[MAX];
} Queue;

int queue[MAX];

// initialize queue
void initQueue(Queue *q) {
  q->front = -1;
  q->rear = -1;
}

// insert element in queue
// Reminder = Dividend - (Divisor * quotient)
void enqueue(Queue *q, int data) {
  // assuming rear = 9 then
  // 9 + 1 % 10 = 10 % 10 = 0
  // which equals to front, making the queue
  // circular when full
  if ((q->rear + 1) % MAX == q->front) {
    printf("Queue is full!\n");
    return;
  } else if (q->front == -1 && q->rear == -1) {
    q->front++;
    q->rear++;
    q->queue[q->rear] = data;
    printf("%d enqueue.\n", q->queue[q->rear]);
  } else {
    q->rear = q->rear + 1 % MAX;
    q->queue[q->rear] = data;
    printf("%d enqueue.\n", q->queue[q->rear]);
  }
}

// dequeue elements
void dequeue(Queue *q) {
  if (q->front == -1 && q->rear == -1) {
    printf("List is empty!\n");
    return;
  } else if (q->rear == 0 && q->front == 0) {
    printf("%d dequeued.\n", q->queue[q->front]);
    q->rear = -1;
    q->front = -1;
  } else {
    printf("%d dequeued.\n", q->queue[q->front]);
    q->front = (q->front + 1) % MAX;
  }
}

// display elements
void display(Queue *q) {
  if (q->front == -1 && q->rear == -1) {
    printf("List is empty!\n");
    return;
  }
  int i = q->front;
  printf("Elements in the list: ");
  while (i <= q->rear) {
    printf("%d ", q->queue[i]);
    i++;
  }
  printf("\n");
}

// main
int main(int argc, char *argv[]) {
  Queue q;
  initQueue(&q);
  enqueue(&q, 10);
  enqueue(&q, 20);
  enqueue(&q, 30);
  enqueue(&q, 40);
  enqueue(&q, 50);
  enqueue(&q, 60);
  display(&q);
  dequeue(&q);
  dequeue(&q);
  display(&q);
}
